keras_nlp/tokenizers/sentence_piece_tokenizer.py

```diff
@@ -0,0 +1,175 @@
+            of `model_file` and `model_bytes` must be set.
+        model_bytes: A SentencePiece serialized model byte array. One of
+            `model_file` and `model_bytes` must be set.
+        sequence_length: If set, the output will be converted to a dense
+            tensor and padded/trimmed so all outputs are of sequence_length.
+
+    References:
+        - [Kudo and Richardson, 2018](https://arxiv.org/abs/1808.06226)
+
+    Examples:
+
+    From bytes.
+    ```python
+    # Train a SentencePiece vocabulary.
+    model = io.BytesIO()
+    ds = tf.data.Dataset.from_tensor_slices(["the quick brown fox."])
+    sentencepiece.SentencePieceTrainer.train(
+        sentence_iterator=ds.as_numpy_iterator(),
+        model_writer=model,
+        vocab_size=20,
+    )
+
+    # Tokenize inputs
+    tokenizer = SentencePieceTokenizer(model_bytes=model.getvalue())
+    ds = ds.map(tokenizer)
+    ```
+
+    From a file.
+    ```python
+    # Train a SentencePiece vocabulary and write it to a file.
+    ds = tf.data.Dataset.from_tensor_slices(["the quick brown fox."])
+    with open("model.proto", "wb") as model_file:
+        sentencepiece.SentencePieceTrainer.train(
+            sentence_iterator=ds.as_numpy_iterator(),
+            model_writer=model_file,
+            vocab_size=20,
+        )
+
+    # Tokenize inputs
+    tokenizer = keras_nlp.tokenizers.SentencePieceTokenizer(
+        model_file="model.proto")
+    ds = ds.map(tokenizer)
+    ```
+    """
+
+    def __init__(
+        self,
+        model_file: str = None,
+        model_bytes: bytes = None,
+        sequence_length: int = None,
+        **kwargs,
+    ) -> None:
+        # Check dtype and provide a default.
+        if "dtype" not in kwargs or kwargs["dtype"] is None:
+            kwargs["dtype"] = tf.int32
+        else:
+            dtype = tf.dtypes.as_dtype(kwargs["dtype"])
+            if not dtype.is_integer and dtype != tf.string:
+                raise ValueError(
+                    "Output dtype must be an integer type or a string. "
+                    f"Received: dtype={dtype}"
+                )
+
+        super().__init__(**kwargs)
+
+        if model_file is None and model_bytes is None:
+            raise ValueError(
+                "One of `model_file` or `model_bytes` must be set. "
+                "Received: `model_file=None`, `model_bytes=None`."
+            )
+
+        if model_file is not None:
+            model_bytes = tf.io.gfile.GFile(model_file, "rb").read()
+        if isinstance(model_bytes, str):
+            model_bytes = base64.b64decode(model_bytes)
+
+        self._sentence_piece = tf_text.SentencepieceTokenizer(
+            model=model_bytes,
+            out_type=self.compute_dtype,
+        )
+
+        # Keras cannot serialize a bytestring, so we base64 encode the model
+        # byte array for saving.
+        self.model_bytes = base64.b64encode(model_bytes).decode("ascii")
```

Comment:
Re the previous comment on ascii. 

My confusion is here it is `base64.b64encode(model_bytes).decode("ascii")`, but on line 115 it is `model_bytes = base64.b64decode(model_bytes)`, which does not mention "ascii".  Is this by intention?

Reply:
Ah I see. I don't think it matters. `base64.b64encode` will return a byte array so we have to convert it to a string for serialization. But I believe `base64.b64decode` can handle a string as input.